H-656, H-657: Improve how we fetch users and orgs, show orgs' avatars

CiaranMn opened the pull request:
## 🌟 What is the purpose of this PR?

The feature change this PR introduces is that organization avatars are now shown on the org's profile page, and in the workspace switcher in the top left.

The bulk of the PR is a refactoring of how we fetch users and orgs, to reduce overfetching, by:
1. Exposing the Graph API's query syntax to the frontend so that we can make requests for an org or user with a given shortname, rather than having to fetch all orgs/users and filter them client-side
2. Introduces a hook to fetch a user or org by either shortname or accountId, and deletes various other hooks that were doing similar things
3. Refactors how users and orgs are loaded when the app loads to reduce request size:
   - for the 'all users' and 'all orgs' requests, stop fetching anything linked to them (we want to stop fetching all users and all orgs altogether, later)
   - for the authenticated user, only fetch the orgs they're a member of and no further at first, and fetch more information about the orgs (members, avatar) in a subsequent request, to speed up the initial user load and avoid bringing in too much of the graph in one request
4. Deletes unused context `RouteWorkspaceInfo`

## Pre-Merge Checklist 🚀

### 🚢 Has this modified a publishable library?

<!-- Confirm you have taken the necessary action to record a changeset or publish a change, as appropriate -->
<!-- Tick AT LEAST ONE box and delete the rest. Do not delete this section! see libs/README.md for info on publishing -->

This PR:

- [x] does not modify any publishable blocks or libraries, or modifications do not need publishing

### 📜 Does this require a change to the docs?

<!-- If this adds a user facing feature or modifies how an existing feature is used, it likely needs a docs change. -->
<!-- Tick ONE box and delete the rest. Do not delete this section! -->

The changes in this PR:

- [x] are internal and do not require a docs change

### 🕸️ Does this require a change to the Turbo Graph?

<!-- If this adds or moves an existing package, modifies `scripts` in a `package.json`, it likely needs a turbo graph change. -->
<!-- Tick ONE box and delete the rest. Do not delete this section! -->

The changes in this PR:

- [x] do not affect the execution graph

## ⚠️ Known issues / next steps

<!-- Are there known issues / intentionally omitted functionality? Flag them here to save reviewers doing so -->

1. Requests for a specific user or org by shortname are slow. H-802 tracks optimizations to be implemented this week
2. On the first page load, an org's avatar won't be available for a second and will load in afterwards. To avoid this we'd have to move org avatar fetching back to the server load – we can see how jarring this remains after further optimizations.

## 🛡 What tests cover this?

<!-- What automated tests cover this? Existing ones? New ones? None? -->

- Existing tests cover user login / construction as part of their flow
- No tests for org profile page / workspace switcher

## ❓ How to test this?

<!-- Tell reviewers how they can test the functionality -->

1. Set an org avatar via its profile page, and check it appears in the workspace switcher and on its public page (`/@[shortname]`)
3. Check that other features affected work, e.g. page creation, mentioning users within a page

## 📹 Demo


https://github.com/hashintel/hash/assets/37743469/5f848442-856a-470f-955c-4120e36a5ddc

nonparibus commented:
@CiaranMn **Can we use the same edit/replace icon background that we use on the file upload interface more generally?** This way it'll work better if the avatar color happens to be the same or similar to the edit pencil's.

We can keep the icon as the edit pencil instead of the 'swap' icon used on the file interface, though.

benwerner01 commented on apps/hash-frontend/src/components/hooks/use-orgs-with-links.ts:
```diff
@@ -0,0 +1,82 @@
+    variables: {
+      operation: {
+        multiFilter: {
+          filters: orgAccountGroupIds.map((entityId) => ({
+            field: ["metadata", "recordId", "uuid"],
+            operator: "EQUALS",
+            value: entityId,
```
Should we try and refer to `entityUuid` explicitly in our code? I think it could lead to confusion otherwise.
```suggestion
          filters: orgAccountGroupIds.map((entityUuid) => ({
            field: ["metadata", "recordId", "uuid"],
            operator: "EQUALS",
            value: entityUuid,
```

CiaranMn replied:
Done in https://github.com/hashintel/hash/pull/3191/commits/677831f0abf9e7f3eebb8d6d4243515633702627

benwerner01 commented on apps/hash-frontend/src/lib/user-and-org.ts:
```diff
@@ -66,202 +82,216 @@ export const constructMinimalUser = (params: {
+    imageEntity: Image;
+  };
+  memberships: {
+    linkEntity: Entity<OrgMembershipProperties>;
```
Maybe we should update the `LinkEntity` type to accept a generic as well, so we can use it here:
```suggestion
    linkEntity: LinkEntity<OrgMembershipProperties>;
```

benwerner01 commented on apps/hash-frontend/src/pages/settings/organizations/new/index.page/create-org-form.tsx:
```diff
@@ -28,6 +30,7 @@ export const CreateOrgForm = () => {
     }
 
     await refetchUser();
+    void refetchOrgs();
```
Don't think it makes a huge difference, but I don't think there's a reason `refetchOrgs` has to happen after `refetchUser`:
```suggestion
    await Promise.all([refetchUser, refetchOrgs]);
```
Or if we don't want the redirect to be blocked:
```suggestion
    void refetchUser();
    void refetchOrgs();
```

CiaranMn replied:
Thanks, done in https://github.com/hashintel/hash/pull/3191/commits/ecc9450d89b7cfe08ef8497ad9db9649a4c56967

Since we just fetch the org on that page we don't rely on the user/full orgs list

benwerner01 reviewed (approved):
Thanks Ciaran, these are great improvements. Whilst testing I had one idea for an improvement, perhaps we can introduce a cancel button/icon of some sort when editing an existing image? This could make it more clear to the user that they haven't just deleted their avatar when clicking the `edit` button:

https://github.com/hashintel/hash/assets/42802102/e3e025d5-933a-4e7d-8ddf-05fb2607ab9e

But that shouldn't block merging, the PR looks good to me 👍

CiaranMn commented:
@nonparibus @benwerner01 I've updated the edit button and added a back button in https://github.com/hashintel/hash/pull/3191/commits/a03da6d4b9889d8ef63eddd8e0f6aecee94a367c


https://github.com/hashintel/hash/assets/37743469/b82cb64f-71fc-414f-9c61-0abb86713fab

benwerner01 reviewed (approved):
Thanks Ciaran, looks good to me 👍
